### PR TITLE
New data queue

### PR DESCRIPTION
**Pull request by oliver-walker-hms**

### CODE REVIEW

File: src/com/hms/flexyosisoftconnector/DataPoint.java [hunk 1]
```diff
@@ -16,8 +16,16 @@
...
+   protected long epochTime;
    protected String timestamp;
 
+   public long getEpcohTime() {
+	   return epochTime;
+   }
+   
+   public void setEpochTime(long time) {
+	   epochTime = time;
+   }
```
Comment (alexjhawk) [inline src/com/hms/flexyosisoftconnector/DataPoint.java]: there should probably be a blank line after this method to keep things consistent

File: README.md [hunk 2]
```diff
@@ -47,6 +47,14 @@ The Flexy's clock must be set to Coordinated Universal Time (UTC) in order to po
 
 Using FTP transfer the contents of the osisoft_connector directory to the /usr/ directory of your eWON via FTP.  If successful you should have a css directory, js directory, config.html, ConnectorConfig.json, flexy-osisoft-connector.jar, and jvmrun located in the /usr/ directory of the Flexy.  You must also follow all steps in the Certificates section below.
 
+### Tag Configuration
+
+The connector reads from the tags historical logs to write to OSIsoft. For best results enable the Logging Deadband feature. Again, the historical logs are sent through the connector to OSIsoft. The settings for the historical log deadband is the same as the settings for the OSIsoft connector deadband.
```
Comment (alexjhawk) [inline README.md]: The "again, ..." sentence is redundant and should probably get removed.

Comment (alexjhawk) [inline README.md]: should be "tag's"

Comment (alexjhawk) [inline README.md]: "For best results" should have a comma after, "For best results, ..."

Comment (alexjhawk) [inline README.md]: "...log deadband is the same..." should be "...log deadband are the same..."

File: README.md [hunk 3]
```diff
@@ -47,6 +47,14 @@ The Flexy's clock must be set to Coordinated Universal Time (UTC) in order to po
...
+
+#### Deadband
+
+With a deadband of 1, the tag value will need to change by 1 before being logged. If you have logging deadband set to -1, we recommend you set the time interval to something high. Set the deadband to whatever your lowest unit of interest is.
```
Comment (alexjhawk) [inline README.md]: The first sentence could make it seem like the value has to change by 1, and only 1. I would change it to say "need to change by greater than 1..."

Comment (alexjhawk) [inline README.md]: Where it says "If you have logging deadband set to -1 ...", I would make it include that -1 also means disabled. "If you have logging deadband disabled (set to -1) ..."

File: src/com/hms/flexyosisoftconnector/DataPoster.java [hunk 4]
```diff
@@ -150,6 +150,10 @@ public void run()
 
             if(retval)
             {
+
+               // pop from historical log queue for first tag
+               //((Tag) OSIsoftConfig.tags.get(0)).popNumTagsFromQueue(2);
+
```
Comment (alexjhawk) [inline src/com/hms/flexyosisoftconnector/DataPoster.java]: Is this something that is going to get uncommented and used?

File: src/com/hms/flexyosisoftconnector/Tag.java [hunk 5]
```diff
@@ -87,28 +114,50 @@ public String getTagValue() {
...
+   public void recordTagValue(String time, String value)
    {
+      // convert time string holding a epoch time in seconds to timestamp
+      long epochSeconds = Long.parseLong(time);
+      long epochTime = epochSeconds*1000;
+      SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-MM-ddTHH:mm:ss");
+      String timestamp = dateFormat.format(new Date(epochTime));
+
+      // Logger.LOG_INFO("time in recordVal is "+timestamp);
+      Logger.LOG_INFO(getTagName()+": revording new val: "+value+" time: "+timestamp);
```
Comment (alexjhawk) [inline src/com/hms/flexyosisoftconnector/Tag.java]: should revording be rewording?

File: src/com/hms/flexyosisoftconnector/DataPoster.java (hunk 4 above)
Comment (oliver-walker-hms) [reply src/com/hms/flexyosisoftconnector/DataPoster.java]: missed this when I was deleting comments. should be deleted

File: src/com/hms/flexyosisoftconnector/Tag.java (hunk 5 above)
Comment (oliver-walker-hms) [reply src/com/hms/flexyosisoftconnector/Tag.java]: this should be changed to LOG_DEBUG or deleted, and i meant recording cause it is in the recordValue function

File: src/com/hms/flexyosisoftconnector/Tag.java [hunk 6]
```diff
@@ -36,6 +61,8 @@
 
    public Tag(String tagName, boolean logDuplicates) {
       eWONTagName = tagName;
+      tagDataFileName += tagName+"TagQueue.txt";
```
Comment (alexjhawk) [inline src/com/hms/flexyosisoftconnector/Tag.java]: This is more personal preference, but if you're defining something in the constructor, I don't think it should be defined in the header above. It makes it hard to follow. 

For example, in the header,
private String tagDataFileName;

and in the constructor,
tagDataFileName = "/usr/internalTagQueue/" + tagName + "TagQueue.txt"

File: src/com/hms/flexyosisoftconnector/Tag.java [hunk 7]
```diff
@@ -120,6 +169,248 @@ public void recordTagValue(String time)
...
+            String line = reader.readLine();
+            while (line != null) {
+
+               // there will be just one line, but just save last line.
+               timeInt = line;
+               // Logger.LOG_INFO("line from time file is: "+line);
+               line = reader.readLine();
+            }
+
+         } catch (IOException e) {
```
Comment (alexjhawk) [inline src/com/hms/flexyosisoftconnector/Tag.java]: If an IOException is thrown, it is possible for "-1" in timeInt to be returned to the calling function. Can those functions still continue if this returns "-1". If not, some more error handling around this is needed.

File: src/com/hms/flexyosisoftconnector/Tag.java [hunk 8]
```diff
@@ -120,6 +169,248 @@ public void recordTagValue(String time)
...
+       exportBlockDescriptor();
+
+       // parse the file generated by the EBD call and add all info to array list
+       parseTagFile();
+   }
+
+
+   // this takes in a timeInt retrieved through the EBD commands on historical logs and converts it
+   // to something we can use in a new EBD command as a start or end time
+   private String convertTimeIntHelper(long timeInt) {
```
Comment (alexjhawk) [inline src/com/hms/flexyosisoftconnector/Tag.java]: method brackets don't match the surrounding methods' brackets

File: src/com/hms/flexyosisoftconnector/Tag.java (hunk 7 above)
Comment (oliver-walker-hms) [reply src/com/hms/flexyosisoftconnector/Tag.java]: I added a error logger in the function that sends out an ebd call which relies on this to not be an error value. I also just returned out of the function before the EBD call was made if that error value is returned. This way the historical log queue would never be updated to something unintentional.

File: README.md [hunk 9]
```diff
@@ -47,6 +47,14 @@ The Flexy's clock must be set to Coordinated Universal Time (UTC) in order to po
 
 Using FTP transfer the contents of the osisoft_connector directory to the /usr/ directory of your eWON via FTP.  If successful you should have a css directory, js directory, config.html, ConnectorConfig.json, flexy-osisoft-connector.jar, and jvmrun located in the /usr/ directory of the Flexy.  You must also follow all steps in the Certificates section below.
 
+### Tag's Configuration
+
+The connector reads from the tags historical logs to write to OSIsoft. For best results, enable the Logging Deadband feature. The settings for the historical log deadband are the same as the settings for the OSIsoft connector deadband.
+
+#### Deadband
+
+With a deadband of 1, the tag value will need to change by a value greater than 1 before being logged. If you have logging deadband disabled (set to -1), we recommend you set the time interval to something high. Set the deadband to whatever your lowest unit of interest is.
```
Comment (alexjhawk) [inline README.md]: what is the time interval?

File: src/com/hms/flexyosisoftconnector/DataPoint.java [hunk 10]
```diff
@@ -16,8 +16,17 @@
    public static final byte TYPE_INT     = 2;
    public static final byte TYPE_DWORD   = 3;
 
+   protected long epochTime;
    protected String timestamp;
 
+   public long getEpcohTime() {
```
Comment (alexjhawk) [inline src/com/hms/flexyosisoftconnector/DataPoint.java]: This should probably be getEpochTime()

File: src/com/hms/flexyosisoftconnector/DataPoster.java [hunk 11]
```diff
@@ -150,7 +150,6 @@ public void run()
 
             if(retval)
             {
-               //Remove all the points that were posted
```
Comment (alexjhawk) [inline src/com/hms/flexyosisoftconnector/DataPoster.java]: should not get deleted

File: src/com/hms/flexyosisoftconnector/Tag.java [hunk 12]
```diff
@@ -36,6 +62,8 @@
 
    public Tag(String tagName, boolean logDuplicates) {
```
Comment (alexjhawk) [inline src/com/hms/flexyosisoftconnector/Tag.java]: mismatched bracket formatting

File: src/com/hms/flexyosisoftconnector/Tag.java [hunk 13]
```diff
@@ -87,28 +115,50 @@ public String getTagValue() {
...
+
+      // parse value into correct type from string and put it in datapoint
       if(validTag)
       {
          DataPoint point = null;
          switch (dataType) {
             case DataPoint.TYPE_BOOLEAN:
-               boolean val = (tagControl.getTagValueAsLong() != 0);
+               boolean val = (Long.parseLong(value) != 0);
                point = new DataPointBoolean(val, time);
```
Comment (alexjhawk) [inline src/com/hms/flexyosisoftconnector/Tag.java]: should time here get changed to timestamp?

File: src/com/hms/flexyosisoftconnector/Tag.java [hunk 14]
```diff
@@ -87,28 +115,50 @@ public String getTagValue() {
...
+    */
+   public void recordTagValue(String time, String value)
    {
+      // convert time string holding a epoch time in seconds to timestamp
+      long epochSeconds = Long.parseLong(time);
+      long epochTime = epochSeconds*1000;
+      SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-MM-ddTHH:mm:ss");
+      String timestamp = dateFormat.format(new Date(epochTime));
+
+      // Logger.LOG_INFO("time in recordVal is "+timestamp);
```
Comment (TomKimsey) [inline src/com/hms/flexyosisoftconnector/Tag.java]: commented out code?

File: src/com/hms/flexyosisoftconnector/Tag.java [hunk 15]
```diff
@@ -87,28 +115,50 @@ public String getTagValue() {
...
+    *
+    * @param time is the time in historical logs in epoch time seconds
+    * (Referred to as timeInt in the EBD generated file)
+    * @param value is the value in historical logs
+    */
+   public void recordTagValue(String time, String value)
    {
+      // convert time string holding a epoch time in seconds to timestamp
+      long epochSeconds = Long.parseLong(time);
+      long epochTime = epochSeconds*1000;
```
Comment (TomKimsey) [inline src/com/hms/flexyosisoftconnector/Tag.java]: Maybe make this more clear that this is milliseconds as you have epochSeconds and epochTime

File: src/com/hms/flexyosisoftconnector/Tag.java [hunk 16]
```diff
@@ -87,28 +115,50 @@ public String getTagValue() {
...
+      SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-MM-ddTHH:mm:ss");
```
Comment (TomKimsey) [inline src/com/hms/flexyosisoftconnector/Tag.java]: can this be static?

File: src/com/hms/flexyosisoftconnector/Tag.java [hunk 17]
```diff
@@ -87,28 +115,50 @@ public String getTagValue() {
...
                break;
             case DataPoint.TYPE_DWORD:
-               point = new DataPointLong(tagControl.getTagValueAsLong(), time);
+               point = new DataPointLong(Long.parseLong(value), timestamp);
                break;
             default:
                return;
          }
+
+         // set the time in epoch time format too for ease of popping off queue
```
Comment (TomKimsey) [inline src/com/hms/flexyosisoftconnector/Tag.java]: too for ease ?

File: src/com/hms/flexyosisoftconnector/Tag.java [hunk 18]
```diff
@@ -120,6 +170,264 @@ public void recordTagValue(String time)
...
+   {
+      // time pointers to keep track of start and end of queue
+      String startTime = "null";
+      String endTime = "null";
+
+      // get the time pointer to use in EBD call for the start of historical logs
+      startTime = readTimePointerFile();
+
+      // check that the error value did not get returned. Log if it did, and do not make EBD call.
+      if (startTime.equalsIgnoreCase("-1")) {
```
Comment (TomKimsey) [inline src/com/hms/flexyosisoftconnector/Tag.java]: -1 should be a defined as a const if it means an error

const int TIME_ERROR = -1;

File: src/com/hms/flexyosisoftconnector/Tag.java [hunk 19]
```diff
@@ -24,6 +38,18 @@
    private List dataPoints;
    private TagControl tagControl;
    private byte dataType;
+   private String tagDataFileName;
+   private String tagTimeIntFileName;
+    private String tagQueueDirectory = "/usr/internalTagQueue/";
```
Comment (TomKimsey) [inline src/com/hms/flexyosisoftconnector/Tag.java]: extra space

Comment (TomKimsey) [inline src/com/hms/flexyosisoftconnector/Tag.java]: should this be const?

Comment (TomKimsey) [reply src/com/hms/flexyosisoftconnector/Tag.java]: and by const I mean final

File: src/com/hms/flexyosisoftconnector/Tag.java [hunk 20]
```diff
@@ -120,6 +170,264 @@ public void recordTagValue(String time)
...
+
+      // end time should be the current time to always keep it up to date
+      endTime = convertTimeIntHelper(System.currentTimeMillis());
+
+      String tagName = getTagName();
+
+      try {
+          // https://ewonsupport.biz/ebd/ <- has a nice EBD string builder. :)
+          Exporter exporter = new Exporter("$dtHL$ftT$st"+startTime+"$et"+endTime+"$fnvalues$tn"+tagName);
+          exporter.ExportTo("file:///"+tagDataFileName);
```
Comment (TomKimsey) [inline src/com/hms/flexyosisoftconnector/Tag.java]: whats up with the triple ///

File: src/com/hms/flexyosisoftconnector/Tag.java [hunk 21]
```diff
@@ -120,6 +170,264 @@ public void recordTagValue(String time)
...
+          exporter.close();
+      }catch (IOException ioe){
+          Logger.LOG_ERR(" error in export block descriptor");
```
Comment (TomKimsey) [inline src/com/hms/flexyosisoftconnector/Tag.java]: space after first "?

File: src/com/hms/flexyosisoftconnector/Tag.java [hunk 22]
```diff
@@ -120,6 +170,264 @@ public void recordTagValue(String time)
...
+               // limit the lines read in to the max queue size to avoid massive queues
+               if(count > MAX_DATA_POINTS_PER_TAG) {
+                   // lines parsed has reached the max, stop parsing.
+                   break;
+               }
+               count++;
+
+               // extract the timeInt, which is first in the file
+               String timeInt = line.substring(0, line.indexOf(';'));
+               //Logger.LOG_INFO(timeInt);
```
Comment (TomKimsey) [inline src/com/hms/flexyosisoftconnector/Tag.java]: commented out code?

File: src/com/hms/flexyosisoftconnector/Tag.java [hunk 23]
```diff
@@ -120,6 +170,264 @@ public void recordTagValue(String time)
...
+    */
+   public void parseTagFile()
+   {
+
+       BufferedReader reader;
+       try {
+           reader = new BufferedReader(new FileReader(tagDataFileName));
+           // read the first line and do nothing with it
+           String line = reader.readLine();
+           if (line != null)
```
Comment (TomKimsey) [inline src/com/hms/flexyosisoftconnector/Tag.java]: in this project and all projects we have tried to avoid multiline conditionals without {}

File: src/com/hms/flexyosisoftconnector/Tag.java [hunk 24]
```diff
@@ -120,6 +170,264 @@ public void recordTagValue(String time)
...
+                * seen. Then it records those indices to use in the substring call.
+                */
+               // main index for the loop
+               int indexMain = 0;
+               // recorded start index of new substring
+               int indexStart = 0;
+               // recorded end index of new substring
+               int indexEnd = 0;
+
+               int firstSemiColonOfInterest = 2;
```
Comment (TomKimsey) [inline src/com/hms/flexyosisoftconnector/Tag.java]: final?

File: src/com/hms/flexyosisoftconnector/Tag.java [hunk 25]
```diff
@@ -120,6 +170,264 @@ public void recordTagValue(String time)
...
+               int secondSemiColonOfInterest = 3;
+               int lastInterestingSemiColon = 4;
```
Comment (TomKimsey) [inline src/com/hms/flexyosisoftconnector/Tag.java]: finals?

File: src/com/hms/flexyosisoftconnector/Tag.java [hunk 26]
```diff
@@ -120,6 +170,264 @@ public void recordTagValue(String time)
...
+               for(int i = 0; i < lastInterestingSemiColon; i++) {
+
+                   // need to add 1 to index main so indexOf function can work
+                   // or it will just find the same thing again
+                   indexMain = 1+ line.indexOf(';', indexMain);
```
Comment (TomKimsey) [inline src/com/hms/flexyosisoftconnector/Tag.java]: formatting is a bit strange, you have no space after the 1 then you have a space after the +

Comment (TomKimsey) [reply src/com/hms/flexyosisoftconnector/Tag.java]: I think they style i was formatting with put spaces around all +, <, ==, etc.

Ex: Logger.LOG_ERR("Tag \"" + tag.getTagName() + "\" does not exist on this eWON");

File: src/com/hms/flexyosisoftconnector/Tag.java [hunk 27]
```diff
@@ -120,6 +170,264 @@ public void recordTagValue(String time)
...
+                   if(i == firstSemiColonOfInterest) {
+                       indexStart = indexMain;
+                   }
+                   if(i == secondSemiColonOfInterest) {
```
Comment (TomKimsey) [inline src/com/hms/flexyosisoftconnector/Tag.java]: why is this not a else if?

File: src/com/hms/flexyosisoftconnector/Tag.java [hunk 28]
```diff
@@ -120,6 +170,264 @@ public void recordTagValue(String time)
...
+          br.write(fileValueString);
+
+      } catch (IOException e) {
+          e.printStackTrace();
+      }finally{
+          try {
+              br.close();
+              fr.close();
+          } catch (IOException e) {
+              e.printStackTrace();
```
Comment (TomKimsey) [inline src/com/hms/flexyosisoftconnector/Tag.java]: no log error here?

File: src/com/hms/flexyosisoftconnector/Tag.java [hunk 29]
```diff
@@ -120,6 +170,264 @@ public void recordTagValue(String time)
...
+      BufferedWriter br = null;
+      try{
+          fr = new FileWriter(file);
+          br = new BufferedWriter(fr);
+
+          // write the data to the file here
+          br.write(fileValueString);
+
+      } catch (IOException e) {
+          e.printStackTrace();
```
Comment (TomKimsey) [inline src/com/hms/flexyosisoftconnector/Tag.java]: no log error here?

File: src/com/hms/flexyosisoftconnector/Tag.java [hunk 30]
```diff
@@ -120,6 +170,264 @@ public void recordTagValue(String time)
...
+              e.printStackTrace();
+          }
+      }
+      return fileValueString;
+   }
+
+   /**
+    * This function is only called when retrieving value for call to export block descriptor.
+    *
+    * the tagTimeIntFileName file holds a time pointer for every tag. 1 file exists per tag.
```
Comment (TomKimsey) [inline src/com/hms/flexyosisoftconnector/Tag.java]: A lot of your comments begin sentences with a lower case but in this case you are switching between the two. I would suggest going through your comments and starting all sentences with capitals.

File: src/com/hms/flexyosisoftconnector/Tag.java [hunk 31]
```diff
@@ -120,6 +170,264 @@ public void recordTagValue(String time)
...
+    * this will hold just a string which will be the epoch time in seconds for where/when
+    * in the Flexy historical logs has the data queue left off on sending data to OSIsoft
+    *
+    * The file can only contain numbers because it is converted to a long
+    *
+    * The below function makes sure the file exists, if not add in the current time to that
+    * otherwise read off the epoch time in that file
+    *
+    * @return Returns a string of the file contents. it will be the epoch time pointer for the data queue.
```
Comment (TomKimsey) [inline src/com/hms/flexyosisoftconnector/Tag.java]: why are you not beginning sentences with capitals?

File: src/com/hms/flexyosisoftconnector/Tag.java [hunk 32]
```diff
@@ -120,6 +170,264 @@ public void recordTagValue(String time)
...
+    * also called a timeInt in the return file from the EBD commands. name will be timeInt throughout code.
+    */
+   public String readTimePointerFile()
+   {
+      String timeInt = "-1";
+      // check if file exists
+      File tempFile = new File(tagTimeIntFileName);
+      boolean exists = tempFile.exists();
```
Comment (TomKimsey) [inline src/com/hms/flexyosisoftconnector/Tag.java]: consider a better name for this boolean, it is not descriptive as to what exists.

File: src/com/hms/flexyosisoftconnector/Tag.java [hunk 33]
```diff
@@ -120,6 +170,264 @@ public void recordTagValue(String time)
...
+         } catch (IOException e) {
+             e.printStackTrace();
+             Logger.LOG_ERR("Something went wrong reading tag data queue time pointer from timeInt file. Check tag timeInt file: "+tagTimeIntFileName);
+         }
+
+      } else { // when connector has not run before, these files will not exist.
+
+          // if file does not exist, set timeInt to the current time to start a new queue
+          Logger.LOG_INFO(getTagName()+": time int was null, setting back of tag data queue current time");
+          long tmpTimeInt = System.currentTimeMillis();
```
Comment (TomKimsey) [inline src/com/hms/flexyosisoftconnector/Tag.java]: units for this variable?

File: src/com/hms/flexyosisoftconnector/Tag.java [hunk 34]
```diff
@@ -120,6 +170,264 @@ public void recordTagValue(String time)
...
+      return formattedTime;
+   }
+
+   /**
+    *  remove from the queue by moving the time pointer forward to a later time
+    * @param numTags number of places to move forward in the queue
+    */
+   public void popNumTagsFromQueue(int numTags)
+   {
+      if (numTags != 0) {
```
Comment (TomKimsey) [inline src/com/hms/flexyosisoftconnector/Tag.java]: what if numTags = -1?

File: README.md [hunk 35]
```diff
@@ -47,6 +47,14 @@ The Flexy's clock must be set to Coordinated Universal Time (UTC) in order to po
 
 Using FTP transfer the contents of the osisoft_connector directory to the /usr/ directory of your eWON via FTP.  If successful you should have a css directory, js directory, config.html, ConnectorConfig.json, flexy-osisoft-connector.jar, and jvmrun located in the /usr/ directory of the Flexy.  You must also follow all steps in the Certificates section below.
 
+### Tag's Configuration
+
+The connector reads from the tag's historical logs to write to OSIsoft. For best results, enable the Logging Deadband feature. The settings for the historical log deadband are the same as the settings for the OSIsoft connector deadband.
+
+#### Deadband
+
+Deadband can be set on the flexy by changing the mode to setup and looking at the historical logging section of the settings. If you set a deadband of 1, then the tag value will need to change by a value of 1 or more before being logged. If you have logging deadband disabled (set to -1) we recommend you set the historical log's logging interval setting to something high if possible. Set the deadband to whatever your lowest unit of interest is.
```
Comment (TomKimsey) [inline README.md]: "changing the mode to setup" what mode? You need to be clear.

Does changing this remove the need for the postduplicatetagvalues parameter? If so that should be removed from the config, webpage, and docs.

File: README.md [hunk 36]
```diff
@@ -47,6 +47,14 @@ The Flexy's clock must be set to Coordinated Universal Time (UTC) in order to po
 
 Using FTP transfer the contents of the osisoft_connector directory to the /usr/ directory of your eWON via FTP.  If successful you should have a css directory, js directory, config.html, ConnectorConfig.json, flexy-osisoft-connector.jar, and jvmrun located in the /usr/ directory of the Flexy.  You must also follow all steps in the Certificates section below.
 
+### Tag's Configuration
+
+The connector reads from the tag's historical logs to write to OSIsoft. For best results, enable the Logging Deadband feature. The settings for the historical log deadband are the same as the settings for the OSIsoft connector deadband.
```
Comment (TomKimsey) [inline README.md]: "The settings for the historical log deadband are the same as the settings for the OSIsoft connector deadband." - what do you mean by this, the osisoft connector doesnt have a deadband setting. Did you mean the historical log deadband setting will be shared with the OSIsoft connector?

File: src/com/hms/flexyosisoftconnector/Tag.java [hunk 37]
```diff
@@ -120,6 +170,264 @@ public void recordTagValue(String time)
...
+    *
+    * The below function makes sure the file exists, if not add in the current time to that
+    * otherwise read off the epoch time in that file
+    *
+    * @return Returns a string of the file contents. it will be the epoch time pointer for the data queue.
+    * also called a timeInt in the return file from the EBD commands. name will be timeInt throughout code.
+    */
+   public String readTimePointerFile()
+   {
+      String timeInt = "-1";
```
Comment (TomKimsey) [inline src/com/hms/flexyosisoftconnector/Tag.java]: what does -1 mean? is it an error or a valid time?

Comment (oliver-walker-hms) [reply src/com/hms/flexyosisoftconnector/Tag.java]: it's an error value

File: src/com/hms/flexyosisoftconnector/DataPoint.java [hunk 38]
```diff
@@ -16,8 +16,19 @@
    public static final byte TYPE_INT     = 2;
```
Comment (TomKimsey) [inline src/com/hms/flexyosisoftconnector/DataPoint.java]: There is a lot of changes in this commit and the commit title is very vague. Ideally a commit like this will have a significant commit message to go along with the title to describe the changes implemented in the commit.

File: src/com/hms/flexyosisoftconnector/Tag.java (hunk 20 above)
Comment (oliver-walker-hms) [reply src/com/hms/flexyosisoftconnector/Tag.java]: That was in the doc's example for using exportTo, so i just tried it that way and it works.. it's 4 slashes with one in the variable. I would guess they use forward slash to escape instead of backslash?

File: src/com/hms/flexyosisoftconnector/Tag.java (hunk 34 above)
Comment (oliver-walker-hms) [reply src/com/hms/flexyosisoftconnector/Tag.java]: I mean, I'll change that to greater than 0 but im pretty sure it is impossible. numTags will only be the pointsAdded from the DataPoster class which is 0 or greater

File: README.md (hunk 36 above)
Comment (oliver-walker-hms) [reply README.md]: I was trying to convey the fact that the historical logs are read into the connector in a clear way.. Probably did the opposite of that. I think I'll just say exactly that - "The historical logs on the Flexy are read into the connector." And I'll change the rest to just tell them to pay attention to the settings for historical logging

File: src/com/hms/flexyosisoftconnector/Tag.java (hunk 34 above)
Comment (TomKimsey) [reply src/com/hms/flexyosisoftconnector/Tag.java]: Just defensive coding.

File: src/com/hms/flexyosisoftconnector/Tag.java (hunk 20 above)
Comment (TomKimsey) [reply src/com/hms/flexyosisoftconnector/Tag.java]: So it does not work with just //?

File: src/com/hms/flexyosisoftconnector/Tag.java (hunk 37 above)
Comment (TomKimsey) [reply src/com/hms/flexyosisoftconnector/Tag.java]: okay, then it should probably be defined like
static final int SOMETHING_ERROR = -1;
String timeInt = SOMETHING_ERROR;

File: src/com/hms/flexyosisoftconnector/Tag.java [hunk 39]
```diff
@@ -120,6 +170,264 @@ public void recordTagValue(String time)
...
 
+   /**
+    * exportBlockDescriptor
+    * send an EBD command with a start and end time and a tag name.
+    * start time is stored in an external file, end time is current time.
+    */
+   public void exportBlockDescriptor()
+   {
+      // time pointers to keep track of start and end of queue
+      String startTime = "null";
```
[the remark this answers is not in the document]
Comment (oliver-walker-hms) [reply src/com/hms/flexyosisoftconnector/Tag.java]: I had planned to do some sort of error checking and forgot. Changing null to TIME_ERROR

File: src/com/hms/flexyosisoftconnector/Tag.java [hunk 40]
```diff
@@ -120,6 +170,264 @@ public void recordTagValue(String time)
...
    }
 
+   /**
+    * exportBlockDescriptor
+    * send an EBD command with a start and end time and a tag name.
+    * start time is stored in an external file, end time is current time.
+    */
+   public void exportBlockDescriptor()
+   {
+      // time pointers to keep track of start and end of queue
```
[the remark this answers is not in the document]
Comment (oliver-walker-hms) [reply src/com/hms/flexyosisoftconnector/Tag.java]: changed to tracker from pointer

File: README.md (hunk 36 above)
Comment (TomKimsey) [reply README.md]: I am not sure if stating just that will be clear, you may want to incorporate that the historical logs are the source for the data pushed to osisoft.

File: src/com/hms/flexyosisoftconnector/Tag.java (hunk 20 above)
Comment (oliver-walker-hms) [reply src/com/hms/flexyosisoftconnector/Tag.java]: looks like it does work with just //
I should of tried that first

Comment (alexjhawk) [reply src/com/hms/flexyosisoftconnector/Tag.java]: Since Flexy supports file://, it shouldn't matter, but there's a difference between file:// and file:///. 

The syntax with file:// should be file://HOSTNAME/path and for file:/// should be file:///path.

file:/// is basically a short version of file://localhost/path, and is meant for localhost file access.

File: README.md (hunk 35 above)
Comment (oliver-walker-hms) [reply README.md]: I guess post duplicate values isn't needed anymore. I'm adding that change to another commit

File: src/com/hms/flexyosisoftconnector/Tag.java [hunk 41]
```diff
@@ -159,7 +153,7 @@ public void recordTagValue(String time, String value)
          // store the time in epoch time format as well so we don't need to convert it back later when popping off queue
          point.setEpochTime(epochTimeMilliseconds);
 
-         if(logDuplicateValues || !point.valueEquals(lastDataPoint))
+         if(!point.valueEquals(lastDataPoint))
```
Comment (TomKimsey) [inline src/com/hms/flexyosisoftconnector/Tag.java]: watch out here - you probably want to remove the if  here.

Comment (oliver-walker-hms) [reply src/com/hms/flexyosisoftconnector/Tag.java]: I'm still concerned that the timestamp would be the same which this would catch. I don't think that would ever be the case but would it hurt to keep this? every 'new' point has a different timestamp

Comment (TomKimsey) [reply src/com/hms/flexyosisoftconnector/Tag.java]: DataPoint's .valueEquals() compares the datapoint's value, not the datapoint itself.

Comment (TomKimsey) [reply src/com/hms/flexyosisoftconnector/Tag.java]: For checking the timestamp equal you would want to look at .equals()

Comment (oliver-walker-hms) [reply src/com/hms/flexyosisoftconnector/Tag.java]: I was looking at the wrong function. I'll change that tomorrow morning

**Comment by TomKimsey** [comment]
Without digging through the functionality myself  will these changes impact the relaying of live data to OSIsoft? In the old data buffer the latest data was sent first, that way OSIsoft is always reporting up to date values and the old values are the least priority. Is that the way it works here or is it posting the oldest points first?

**Comment by oliver-walker-hms** [comment]
it grabs the oldest data first in batches of up to 1000 per tag and tries to feed that into the previous queue system. so it will chunk the data, grab the oldest chunk and do the newest of the chunk first.

**Comment by TomKimsey** [comment]
So that behavior was purposely avoided in the old design, we intentionally designed the system so that the user would always get the freshest data first. That way if something is going wrong you can take action when its going wrong.

**Comment by TomKimsey** [comment]
Old data is given a lower priority to new data.

**Comment by oliver-walker-hms** [comment]
it'll be slow

**Comment by TomKimsey** [comment]
Are you not able to pull from both sides of the logs?

**Comment by oliver-walker-hms** [comment]
i need to sort the generated files myself. seems to be impossible to just swap the sort order by the EBD command.

**Comment by oliver-walker-hms** [comment]
it wont be that much slower but it will add a good amount of time to process it

**Comment by TomKimsey** [comment]
In the EBD command you can specify the start and end time for the historical logs you pull right?

If so you could pull from the old data, then do another EBD command to pull from the new data. 
Ex 400 old points + 100 new points.

**Comment by oliver-walker-hms** [comment]
@Tom-HMS  Tom was okay with the new queue processing older data first

File: README.md [hunk 42]
```diff
@@ -47,6 +47,20 @@ The Flexy's clock must be set to Coordinated Universal Time (UTC) in order to po
...
 
+### Tag's Configuration
+
+The tag's historical logs are the source for the data pushed to OSIsoft, so settings for your historical logging will effect what data can get sent to OSIsoft. Make sure you enable historical logging in your tag settings. 
++ On the flexy web page, navigate to Tags -> values
++ On the web top left section of the page, switch the mode from view to setup
++ Click on your tag that you wish to configure, and click on edit
++ On the tag configuration panel scroll down and enable Historical Logging
+
+We suggest you make use of both the available historical logging settings: Logging Deadband and Logging Interval. More information on Logging DeadBand is listed below.
```
Comment (TomKimsey) [inline README.md]: we - first person

Consider rephrasing to something like
It is suggested to make use ...

File: README.md [hunk 43]
```diff
@@ -47,6 +47,20 @@ The Flexy's clock must be set to Coordinated Universal Time (UTC) in order to po
...
+
+#### Logging Deadband
+
+If you set a deadband of 1, then the tag value will need to change by a value of 1 or more before being logged. If Deadband is set to 2, the value needs to change by 2 or more before being logged. If you have logging deadband disabled (set to -1) we recommend you set the historical log's logging interval setting to something high if possible. Set the deadband to whatever your lowest unit of interest is.
```
Comment (TomKimsey) [inline README.md]: We used again

I would even consider getting rid of the second person use cases in this pull request. This readme is not completely third person but it might be good to reword it to be.

File: src/com/hms/flexyosisoftconnector/Tag.java [hunk 44]
```diff
@@ -120,6 +170,265 @@ public void recordTagValue(String time)
...
+      endTime = convertTimeIntHelper(System.currentTimeMillis());
+
+      String tagName = getTagName();
+
+      try {
+          // https://ewonsupport.biz/ebd/ <- has a nice EBD string builder. :)
+          Exporter exporter = new Exporter("$dtHL$ftT$st" + startTime + "$et" + endTime + "$fnvalues$tn" + tagName);
+          exporter.ExportTo("file://" + tagDataFileName);
+          exporter.close();
+      }catch (IOException ioe){
```
Comment (TomKimsey) [inline src/com/hms/flexyosisoftconnector/Tag.java]: no space after } and no space after {

File: src/com/hms/flexyosisoftconnector/Tag.java [hunk 45]
```diff
@@ -120,6 +170,265 @@ public void recordTagValue(String time)
       }
    }
 
+   /**
+    * exportBlockDescriptor
+    * Send an EBD command with a start and end time and a tag name.
+    * Start time is stored in an external file, end time is current time.
+    */
+   public void exportBlockDescriptor()
```
Comment (TomKimsey) [inline src/com/hms/flexyosisoftconnector/Tag.java]: Should this be names something else? It seems to be a very specific EBD request and not a generic method for all EBDs.

File: src/com/hms/flexyosisoftconnector/Tag.java [hunk 46]
```diff
@@ -120,6 +164,265 @@ public void recordTagValue(String time)
...
+                   // Convert last point's epoch time to seconds, which is what the EBD "timeInt" is in
+                   long lastPointSeconds = lastDataPoint.epochTime/1000;
+                   if (lastPointSeconds != Long.parseLong(timeInt)) {
+                       recordTagValue(timeInt, value);
+                   }
+               }
+
+               // Read next line, repeat
+               line = reader.readLine();
+           }
```
Comment (alexjhawk) [inline src/com/hms/flexyosisoftconnector/Tag.java]: There should be a call made to `reader.close()` at the end of this try

File: src/com/hms/flexyosisoftconnector/Tag.java [hunk 47]
```diff
@@ -120,6 +164,265 @@ public void recordTagValue(String time)
...
+         try {
+            reader = new BufferedReader(new FileReader(tagTimeIntFileName));
+            String line = reader.readLine();
+            while (line != null) {
+
+               // There will be just one line, but just save the last line.
+               timeInt = line;
+               line = reader.readLine();
+            }
+
```
Comment (alexjhawk) [inline src/com/hms/flexyosisoftconnector/Tag.java]: Missing a call to `reader.close()`
